### DISCUSSION TITLE
Not seeing benefits of startTransition compared to web

### DISCUSSION THREAD

**Discussion by Almouro** [reactwg/react-native-new-architecture#94] [Q&A]
Hi, 

I was trying to play with the new architecture and react 18 concurrent features, namely `startTransition` however, weirdly I'm not seeing any benefit, while the same code works really well on the web.

Apologies if I've missed something completely obvious! I did see `Running "FabricEnabled" with {"fabric":true,"initialProps":{"concurrentRoot":true},"rootTag":1}` in my packager logs.

Here's the same code running on a release APK ([available here](https://github.com/Almouro/fabric-test/blob/add-concurrent-scenario/apks/fabric-concurrent.apk) and [a react-native-web version](https://keen-kulfi-b15db1.netlify.app/). 
Both on the same device (high end Android device Samsung S10 Android 12)

- the green part is 30k views to simulate a slow UI
- On the left, we don't see any improvement on the App
- On the right, we can clearly see that it's so much better at updating the top value in priority

https://user-images.githubusercontent.com/4534323/200001450-5a49a5b5-b2dd-403f-a75a-936cba49c1b8.mp4

Here's the code:
```ts
const SlowUI = React.memo(({value}: {value: number}) => (
  <>
    <Text style={styles.text}>SLOW UI Value: {value}</Text>
    <View style={{flexDirection: 'row', flexWrap: 'wrap'}}>
      {Array(30000)
        .fill(1)
        .map((_, index) => (
          <View
            key={index}
            style={{width: 1, height: 1, backgroundColor: 'green'}}
          />
        ))}
    </View>
  </>
));

const WithStartTransition = () => {
  const [value, setValue] = React.useState(0);
  const [deferredValue, setDeferredValue] = React.useState(0);
  const [isPending, startTransition] = React.useTransition();

  const handleClick = () => {
    const newValue = value + 1;
    setValue(newValue);
    startTransition(() => {
      setDeferredValue(newValue);
    });
  };

  return (
    <>
      <Button onPress={handleClick} title="Increment value" />
      <Text style={styles.text}>Value: {value}</Text>
      <Text style={styles.text}>
        DeferredValue: {deferredValue} {isPending ? 'PENDING' : ''}
      </Text>
      <SlowUI value={deferredValue} />
    </>
  );
};
```

**Comment by hurali97** [comment]
Hi everyone adding to @Almouro's findings, 

I was experimenting with React18 features like `startTransition`, `useDeferredValue` and `suspense` in react native. In my findings, I have pretty much unexpected results. Here’s my [repo](https://github.com/hurali97/react18features) where I have created 2 apps, one in  react and other in react-native. So in terms of features we should have parity b/w both of them.

**startTransition**:
- This works same on both react and react-native. The update that’s wrapped in `startTransition` results in `isPending` being set to true when React is rendering the UI and decides to flush the urgent one first. Then it renders the non-urgent one and toggles the `isPending` to false. 🟢

**useDeferredValue**:
- This one behaves different on react & RN. On react, when we type in input we render a very slow list based on the input state value and without using uDV, we can hardly type in the input. Since, react tries to render both of them. So instead we can tell React to defer rendering until urgent one is flushed. For this, we now pass the input state value to this uDV hook and pass the returned value to Slow UI component and the difference is so smooth. On the other hand, I am unable to notice any difference with or without uDV on react native. 🔴

<details>
<summary>React</summary>

https://user-images.githubusercontent.com/47336142/210064309-d76bf80b-c94f-45c1-8d8d-fa350694e4c8.mp4

</details>

<details>
<summary>React Native using uDV</summary>

https://user-images.githubusercontent.com/47336142/210072435-08914b95-23cb-4581-8fdc-ec6ed09a2040.mp4

</details>

<details>
<summary>React Native not using uDV</summary>

https://user-images.githubusercontent.com/47336142/210072478-04876588-31a0-47e2-8782-847092b54c1f.mp4

</details>

Inspiration for the above example took from this [sandbox](https://codesandbox.io/s/jovial-frost-30embe) available on react's website

**suspense with startTransition**:
- On react, `suspense` works with Relay like models or with react.lazy. So I tried with both of them and `startTransition` integrates very well with `suspense`. When we use `suspense` without `startTransition` we will see the fallback components. Which may seem not good at times, where we want the user to be able to look at the old content while new one is being prepared. So we can integrate `startTransition` to help us in that. Updates wrapped in `startTransition` will not show the fallback component, which will allow to show the old UI to the user. In our example, we notice that if we don’t use `startTransition`, fallback of **_Loading the app..._** will be displayed. If we use `startTransition`, it won’t be shown. It works well on react, but on react-native it doesn’t work as intended. We still get Loading the app displayed even after using `startTransition`. 🔴

**Note the **_Loading the app..._** fallback component that is not visible in React but is visible in React Native**

<details>
<summary>React</summary>

https://user-images.githubusercontent.com/47336142/210064437-3ec74be2-d23b-4102-8b52-e5dbf20e2067.mp4

</details>


<details>
<summary>React Native</summary>

https://user-images.githubusercontent.com/47336142/210071834-63fcda7f-22fa-4c9f-a28d-edfc9eeca7d8.mp4

</details>

Inspiration for the above example took from this [sandbox](https://codesandbox.io/s/nice-shaw-k6t10j?file=/src/index.js) available on react's website


More resources:
- https://beta-reactjs-org-git-api-suspense-2-fbopensource.vercel.app/apis/react/useDeferredValue
- https://17.reactjs.org/docs/concurrent-mode-patterns.html

**Comment by joshuayoes** [comment]
Adding to this discussion: as of `react-native@0.72.0-rc.1`, `useTransition` appears to work as expected on Android but not iOS. 

I've created a reproduction repository here: https://github.com/joshuayoes/use-transition-react-native-example

This repository implements the [React docs example of the useTransition hook](https://react.dev/reference/react/useTransition#usage) in React Native on version 0.72.0-rc.1.

This demo tap the Posts (slow) button and then the Contacts button. According the docs, tapping the Contacts button should show the ContactTab view on the next render and skip committing the render of the PostsTab view.

On Web and Android, only 2 renders happen, skipping the slow `PostsTab` render tree. However, on iOS, 4 renders happen, synchronously blocking the next render until the slow `PostsTab` render tree is commited.

# Demos
| Web | Android | iOS |
|--------|--------|--------|
| <video src="https://user-images.githubusercontent.com/37849890/235544806-da8a2a2f-b11d-46b8-aa8d-d38dd954734f.mov" /> |  <video src="https://user-images.githubusercontent.com/37849890/235544680-33106362-de39-42fb-b3aa-9b7ac5ee13b8.mov" /> |  <video src="https://user-images.githubusercontent.com/37849890/235544732-dbf3b2f9-ffb2-4f00-87bc-6a592e110d79.mov" />  | 

# Profiles

## Android

Using `npx react-devtools`, you can see that Android only renders twice. 

1. Initial Render
2. `ContactTab` render

The "Posts (slow)" view is not committed to the render tree because the "Contact" view finishes rendering first. [See React DevTools profile export for more information.](https://github.com/joshuayoes/use-transition-react-native-example/files/11368534/android-profile.json.zip)


<img width="800" alt="profile-android-1" src="https://user-images.githubusercontent.com/37849890/235545127-bb1a6af0-830f-4dc9-8e0d-1dbd0349e058.png">
<img width="800" alt="profile-android-2" src="https://user-images.githubusercontent.com/37849890/235545132-3fb83563-3c45-4aae-a7f9-e4a938f515af.png">

## iOS

On iOS, this example shows 4 renders. 

1. Initial Render
2. `PostsTab` render
3. `onClick` props change to `TabButton`
4. `ContactTab` render

 [See React DevTools profile export for more information.](https://github.com/joshuayoes/use-transition-react-native-example/files/11368552/ios-profile.json.zip)

<img width="800" alt="profile-ios-1" src="https://user-images.githubusercontent.com/37849890/235545490-a65de350-0560-4de4-af09-66f4cf92fa4c.png">
<img width="800" alt="profile-ios-2" src="https://user-images.githubusercontent.com/37849890/235545446-22386598-d766-4db7-8af2-e9044edd4a93.png">
<img width="800" alt="profile-ios-3" src="https://user-images.githubusercontent.com/37849890/235545449-19877615-d3db-4941-97ad-fdd2b99f249b.png">
<img width="800" alt="profile-ios-4" src="https://user-images.githubusercontent.com/37849890/235545450-8ba3b877-0b0a-428a-995d-10dd3eda66a0.png">

**Reply by joshuayoes** [reply]
Update for `react-native@0.72.0-rc.3`, useTransition now works on iOS but Android has regressed to previous unexpected synchronous rendering behavior. 

See commit with upgrade: https://github.com/joshuayoes/use-transition-react-native-example/commit/ae0756246e80bcfbc6ea8386f0cfedf7ea227d1a

# Demos
| Web | Android | iOS |
|--------|--------|--------|
| <video src="https://user-images.githubusercontent.com/37849890/235544806-da8a2a2f-b11d-46b8-aa8d-d38dd954734f.mov" /> |  <video src="https://github.com/reactwg/react-native-new-architecture/assets/37849890/29320733-3c02-479d-9ab0-20d9300a7b68" /> |  <video src="https://github.com/reactwg/react-native-new-architecture/assets/37849890/09c13b98-baaa-4dfb-a152-cb15113275b8" />  | 

# Profiles

## Android

[See React DevTools profile export for more information.](https://github.com/reactwg/react-native-new-architecture/files/11459250/0.72.0-rc.3-android.json.zip)

<img width="500" alt="0 72 0-rc 3-android-1" src="https://github.com/reactwg/react-native-new-architecture/assets/37849890/d920d6d6-e5cc-4054-a1db-db46d7b36c22">
<img width="500" alt="0 72 0-rc 3-android-2" src="https://github.com/reactwg/react-native-new-architecture/assets/37849890/5075f53d-84c4-44b5-af21-61eed4786a73">
<img width="500" alt="0 72 0-rc 3-android-3" src="https://github.com/reactwg/react-native-new-architecture/assets/37849890/d47ad18f-8b2d-4263-9ae9-7428ab0d42c4">
<img width="500" alt="0 72 0-rc 3-android-4" src="https://github.com/reactwg/react-native-new-architecture/assets/37849890/46ba0e0f-61d1-4589-b625-79d276122bc5">

## iOS

[See React DevTools profile export for more information.](https://github.com/reactwg/react-native-new-architecture/files/11459260/0.72.0-rc.3-ios.json.zip)

<img width="500" alt="0 72 0-rc 3-ios-1" src="https://github.com/reactwg/react-native-new-architecture/assets/37849890/8fa80f75-cf86-419e-a369-de405dad17ee">
<img width="500" alt="0 72 0-rc 3-ios-2" src="https://github.com/reactwg/react-native-new-architecture/assets/37849890/a3f44247-841b-4235-bcf2-681c3196a9a5">

**Comment by cipolleschi** [comment]
Sorry for the very late answer.

@Almouro I was trying your reproducer with RN 0.79.3. I think the situation is much better, especially for Android in Release mode.
| Android (Debug) | Android  (Release) | iOS (Debug) |
| --- | --- | --- |
| <img src="https://github.com/user-attachments/assets/f42539f5-0e64-4252-9fcd-930ad865e006" height="480" /> |  <img src="https://github.com/user-attachments/assets/250d7640-577a-4d51-a1bb-05d49287a860" height="480" /> | <img src="https://github.com/user-attachments/assets/15deb364-6161-405f-a603-e8b76d7a43a6" height="480" /> |

---

@hurali97 I tried yours as well, and effectively, I can't see any difference when using deferred value... I'll try to raise it internally.

--- 

@joshuayoes I tried the reproducer here too.
I think this is now fixed in 0.79.3

| Android | iOS |
| --- | --- |
| <img src="https://github.com/user-attachments/assets/b60f982b-4bf3-445b-a164-59c4a317aa98" height="480" /> | <img src="https://github.com/user-attachments/assets/b1959079-da08-430b-beb2-8c1363ce5b7f" height="480" /> |
